lib/wherebot.rb

```diff
@@ -8,34 +8,61 @@ class Wherebot
 
   class << self
     def update_wheres(destructive: false)
-      imap = Net::IMAP.new(
-        ENV['INCOMING_MAIL_SERVER'], ENV['INCOMING_MAIL_PORT'], true
-      )
+      @destructive = destructive
+      open_imap
+      sign_into_wheremail
+      import_emails
+      clean_up
+    end
 
-      sign_into_wheremail(imap)
+    def forget_old_messages
+      old = WhereMessage.where('sent_at < ?', 30.days.ago)
 
-      imap.search(["UNSEEN"]).each do |id|
-        Wherebot::Message.new(imap, id, destructive).create
+      if old.present?
+        puts "destroying #{old.count} old messages."
+      else
+        puts "no old messages."
       end
 
-      clean_up(imap, destructive)
+      old.destroy_all
     end
 
-    def forget_old_messages
-      WhereMessage.where('sent_at < ?', 30.days.ago).destroy_all
+    private
+
+    def open_imap
+      @imap = Net::IMAP.new(
+        ENV['INCOMING_MAIL_SERVER'], ENV['INCOMING_MAIL_PORT'], true
+      )
+      puts "connected to IMAP" if @imap.is_a?(Net::IMAP)
+      @imap
     end
```

Comment:
An alternative could be to
```
def imap
  @imap ||= Net::IMAP.new(
    ENV['INCOMING_MAIL_SERVER'], ENV['INCOMING_MAIL_PORT'], true
  )
end
```
Then you can call a cached version of the imap object without having to make sure that you've initialized it by calling `open_imap`.

Not blocking and otherwise this looks good to me.

Reply:
I like this approach in general, but in this case the variable is created in a class method, rather than an instance method, so if we memoize the object it'll just hang out forever, getting more and more broken.  It's important to instantiate a new instance of IMAP every time the method is called.

An alternative is to void `@imap` as part of the cleanup step, but i don't think the cleanup step should know that much about the implementation of the imap creation step.

I've renamed the method `create_new_imap` to try to convey that nuance.

Reply:
Whoa, i just got a flash of all the instances i've created and never cleaned up, because Rails doesn't make you clean things up.  What happened to them?  Are they all still in memory somewhere?